src/shaders/_prelude.glsl

```diff
@@ -0,0 +1,44 @@
+// IMPORTANT:
+// This prelude is injected in both vertex and fragment shader be wary
+// of precision qualifiers as vertex and fragment precision may differ
+
+#define EPSILON 0.0000001
+#define PI 3.141592653589793
+
+#ifdef FOG
+
+uniform mediump vec4 u_fog_color;
+uniform mediump vec2 u_fog_range;
+uniform mediump float u_fog_horizon_blend;
+
+varying vec3 v_fog_pos;
+
+float fog_range(float depth) {
+    // Map [near, far] to [0, 1] without clamping
+    return (depth - u_fog_range[0]) / (u_fog_range[1] - u_fog_range[0]);
+}
+
+// Assumes z up and camera_dir *normalized* (to avoid computing
+// its length multiple times for different functions).
+float fog_horizon_blending(vec3 camera_dir) {
+    float t = max(0.0, camera_dir.z / u_fog_horizon_blend);
+    // Factor of 3 chosen to roughly match smoothstep.
+    // See: https://www.desmos.com/calculator/pub31lvshf
```

Comment:
I imagine these desmos links as being a nice convenience but not essential to understanding the code, but maybe it'd be a good idea for me to screenshot them and add them to the github PR before this work is all water under the bridge.